admin/class-smaily-for-wp-admin.php

```diff
@@ -67,6 +67,57 @@ public function enqueue_scripts() {
 		}
 	}
 
+	/**
+	 * Verify plugin was updated via transient and trigger DB upgrades.
+	 *
+	 * @since 3.0.0
+	 */
+	public function listen_for_upgrade_transient() {
+		if ( get_transient( 'smailyforwp_plugin_updated' ) !== true ) {
+			return;
+		}
+
+		$plugin_version = SMLY4WP_PLUGIN_VERSION;
+		if ( $plugin_version === get_option( 'smailyforwp_db_version' ) ) {
+			return delete_transient( 'smailyforwp_plugin_updated' );
+		}
+
+		if ( version_compare( $plugin_version, '3.0.0', '=' ) ) {
+			require_once SMLY4WP_PLUGIN_PATH . '/migrations/upgrade-3-0-0.php';
+			smailyforwp_upgrade_3_0_0();
+		}
```

Comment:
This code is duplicated between activation and upgrade.

I suggest you create a new lifecycle class, that takes care of plugin activation, upgrade and uninstall logic.

You could use pattern:
```php
class Lifecycle {
    public function activate() {
        $this->run_migrations();
    }

    public function upgrade() {
        if ( ! get_transient() !== true ) {
            return;
        }

        $this->run_migrations();

        delete_transient();
    }

    public function uninstall() {
        // ...
    }

    public function check_for_upgrade() {
        // ...
    }

    private function run_migrations() {
        require 'migration.php';
        run_migration_to_version();
    }
}
```

That can be used through plugin class:
```php
private function define_lifecycle_hooks() {
    $plugin_lifecycle = new Lifecycle();

    register_activation_hook( SMLY4WP_PLUGIN_FILE, array( $plugin_lifecycle, 'activate' ) );
    register_uninstall_hook( SMLY4WP_PLUGIN_FILE, array( $plugin_lifecycle, 'uninstall' ) );

    $this->loader->add_action( 'plugins_loaded', $plugin_lifecycle, 'upgrade' );
    $this->loader->add_action( 'upgrader_process_complete', $plugin_lifecycle, 'check_for_upgrade', 10, 2 );
}
```

This way you can get proper boundaries between classes and can make plugin initialization much simpler:
```php
require SMLY4WP_PLUGIN_PATH . '/includes/class-smaily-for-wp.php';

function run_smaily_for_wp() {
	$plugin = new Smaily_For_WP();
	$plugin->run();
}
run_smaily_for_wp();
```

Reply:
Added the lifecycle class. It really helped make the upgrade process clearer and simpler.